binlog.cpp

```diff
@@ -0,0 +1,180 @@
+        return buffer->len;
+    }
+
+    uint8_t *payload;
+
+    if (mavlink2) {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+        trimmed_zeros = get_trimmed_zeros(buffer);
+    } else {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+        trimmed_zeros = 0;
+    }
+
+    if (trimmed_zeros) {
+        binlog_data
+            = (mavlink_remote_log_data_block_t *)alloca(sizeof(mavlink_remote_log_data_block_t));
+        memcpy(binlog_data, payload, payload_len);
+        memset(binlog_data + payload_len, 0, trimmed_zeros);
+    } else {
+        binlog_data = (mavlink_remote_log_data_block_t *)payload;
+    }
+
+    if (_logging_start_timeout) {
+        // If message doesn't start a log session, just ignore it
+        if (binlog_data->seqno == 0) {
+            _remove_start_timeout();
+            if (!_start_alive_timeout()) {
+                log_warning("Could not start liveness timeout - mavlink router log won't be able "
+                            "to detect if flight stack stopped");
+            }
+        } else {
+            return buffer->len;
+        }
+    }
+    _logging_data_process(binlog_data);
+
+    _stat.write.total++;
+    _stat.write.bytes += buffer->len;
+
+    return buffer->len;
+}
+
+void BinLog::_send_ack(uint32_t seqno)
+{
+    mavlink_message_t msg;
+
+    // Message filled a gap, or is duplicated. just send the ack
+    if (seqno < _last_acked_seqno) {
+        mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg,
+                                                 _target_system_id, MAV_COMP_ID_ALL, seqno, 1);
+        _send_msg(&msg, _target_system_id);
+        return;
+    }
+
+    // TODO maybe a threshould of [n]acks to be sent?
+    // TODO send ack to source only?
+    // Send nacks regarding unseen seqno
+    for (uint32_t i = _last_acked_seqno; i < seqno; i++) {
+        mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg,
+                                                 _target_system_id, MAV_COMP_ID_ALL, seqno, 0);
+        _send_msg(&msg, _target_system_id);
+    }
+
+    // Send ack to seen seqno
+    mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg, _target_system_id,
+                                             MAV_COMP_ID_ALL, seqno, 1);
+    _send_msg(&msg, _target_system_id);
+    _last_acked_seqno = seqno;
+}
+
+void BinLog::_logging_data_process(mavlink_remote_log_data_block_t *msg)
+{
+    ssize_t r;
+
+    if (lseek(_file, msg->seqno * MAVLINK_MSG_REMOTE_LOG_DATA_BLOCK_FIELD_DATA_LEN, SEEK_SET) < 0) {
+        log_error_errno(errno, "lseek failed (%m)");
+        return;
+    }
+
+    r = write(_file, msg->data, MAVLINK_MSG_REMOTE_LOG_DATA_BLOCK_FIELD_DATA_LEN);
+    if (r != MAVLINK_MSG_REMOTE_LOG_DATA_BLOCK_FIELD_DATA_LEN) {
+        // partial writes are handled by not sending ack. Flight stack should resend
```

Comment:
if `r == -1` then you should do something more serious than logging an error.  There's a good chance that that log message will be lost because you've just filled the disk up!